### PR TITLE
Fix `numpy.dtype` hashing for `numpy >= 1.20`

### LINKED ISSUE #1080: Hashing : possible work-around for avoiding to rely on `np.dtype('float64').__class__` being picklable
From a failure in the scikit-learn tests using the numpy development wheels (1.20.dev) https://github.com/scikit-learn/scikit-learn/pull/17707#issuecomment-650212919.

The error is related to `joblib.Memory` argument hashing and its special dtype treatment:
https://github.com/joblib/joblib/blob/ac0f1528ffec9cd53cabbafaca887ac880697862/joblib/hashing.py#L222-L235

This comes from the fact that `np.dtype('float64').__class__` is not picklable in numpy dev:

```py
import pickle

import numpy as np


print('numpy version:', np.__version__)
dt = np.dtype('float64')
print('class: ', dt.__class__)
try:
    pickle.dumps(dt.__class__)
    print('Can pickle dt.__class__')
except Exception as exc:
    print('Exception while pickling dt.__class__')
    print(exc)
```

Output numpy-dev:
```
numpy version: 1.20.0.dev0+be8ab91
class:  <class 'numpy.dtype[float64]'>
Exception while pickling dt.__class__
Can't pickle <class 'numpy.dtype[float64]'>: attribute lookup dtype[float64] on numpy failed
```

Ouptut numpy 1.19.0:
```
numpy version: 1.19.0
class:  <class 'numpy.dtype'>
Can pickle dt.__class__
```

I opened https://github.com/numpy/numpy/issues/16692 to discuss whether this could be fixed in numpy.

Maybe we could use something else in the hash (for example the fully qualified class name of the dtype? Maybe `.descr` is enough actually?). One annoying side-effect is that this would change the hash and cause recomputations for `joblib.Memory` caches when you update numpy. At the same time the changes in numpy dtype (in numpy master) will very likely cause the same recomputations anyway ...

cc @ogrisel, @pierreglaser

Context: #239 #251

### PR DESCRIPTION
**Pull request by pierreglaser**
In order to fix #1080, where our old way of hashing a `numpy.dtype` was broken by recent changes in `numpy.dtype` implementation, I propose to rely on  loss-less, standard pickling to hash `dtypes`. The risk, (which is the reason, I assume, it was not done this way already) is that the `pickle` memoization process will interfere will hashing and create spurious changes in  `pickle` string of `dtypes` with the final consequence of assigning different hash values for seemingly identical objects (see more detail in the comments that I wrote in the code introduced by this PR). To short-circuit memoization, I propose to make a deepcopy of each type that is hashed prior to hashing it.

Note: this change in `dtype` hashing makes that in the future, when users will upgrade their   `joblib` to a release that include this code, all hashes (and thus `joblib` caches) generated using previous `version` will be invalidated. This conflicts with a  semi-public contract of joblib (written in the comments of tests, at least) that upgrades of joblib should not interfere with caching.

We should circumvent this issue by including this change in `joblib 1.0` (note the major version bump), and not before. This should happen soon though, as `numpy 1.20` is planned to be released in the near future.

In the future,  `joblib` should not claim to guarantee hashing consistency across environment changes (be the change related to `joblib`, or any other library).

TODO:

- [x] make a mention of how `joblib` does not guarantee hashing consistency/cache validation as soon as the working environment is altered (update of either joblib, or any other library)
- [x] update `CHANGES`

cc @ogrisel. Still a bit wip, but first batch of comments welcomed.

### CODE REVIEW

File: joblib/hashing.py
Comment (pierreglaser) [inline joblib/hashing.py:235]: @ogrisel turns out `deepcopy` does not make my test pass, but `pickle.loads(pickle.dumps(obj))` does. I made some pickle string introspection and it seems related to  `string` interning, something that popped up very early in the `joblib`/ `dtype` history (cant retrieve a source for now),

File: joblib/test/test_hashing.py
Comment (pierreglaser) [inline joblib/test/test_hashing.py:492]: (Leaving a comment to make sure this line has enough visibility)

**Comment by pierreglaser** [comment]
(rebased)

File: CHANGES.rst
Comment (ogrisel) [inline CHANGES.rst:7]: Make `joblib.hash` and `joblib.Memory` caching system compatible...

File: CHANGES.rst
Comment (ogrisel) [inline CHANGES.rst:9]:
```suggestion
  fix will see their previous cache invalidated if they
```

The cache is not thus wiped out, the old entries are still on the disk.

File: doc/memory.rst
Comment (ogrisel) [inline doc/memory.rst:391]: same remark: invalidated instead of wiped out

File: doc/memory.rst
Comment (ogrisel) [inline doc/memory.rst:403]: Since we have not easy to maintain way to ensure this contract (w.r.t. versioning semantics) I would rather just loosen the contract such that:


The values returned by `joblib.hash` are not guaranteed to stay constant when upgrading joblib to a new version and therefore the entries of a `joblib.Memory` cache can also get invalidated when upgrading as a result. Invalidation can also happen when the parameters of a cached function are instances of a classes defined in upgraded third party libraries such as numpy for instance.

File: joblib/hashing.py
Comment (ogrisel) [reply joblib/hashing.py:236]: Alright...

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:478]: Put this code in a private helper function and call it twice to get the 2 lists of objects instead.

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:490]:
```suggestion
        # Making sure memoization and object identity do not interfere with
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:488]: This an interesting test, but I would first test the following cases:

```python
        e1 = ProcessPoolExecutor(max_workers=1)
        e2 = ProcessPoolExecutor(max_workers=1)

        hash_1 = e1.submit(hash, obj_1).result()
        hash_2 = e2.submit(hash, obj_1).result()
        assert hash_1 == hash_2

        hash_3 = e1.submit(hash, obj_2).result()
        assert hash_1 == hash_3
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [reply joblib/test/test_hashing.py:492]: what do you mean by joblib master? joblib 0.17?

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:495]: Would the following assertion pass?:

```python
assert hash([obj_1, obj_1]) == hash([obj_1, obj_2])
```

maybe we have another memoization bug related to string interning.

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:417]:
```suggestion
    # hash of an object with some hard-coded target value to guarantee that
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:480]: I think we should use a context manager to clean the executor(s) properly before the end of the test.

**Review by ogrisel — COMMENTED** [review]
First pass of review.

File: joblib/hashing.py
Comment (ogrisel) [inline joblib/hashing.py:237]: To avoid re-parsing the pickle, we could directly do:

```python
self._hash.update("_HASHED_DTYPE")
self._hash.update(pickle.dumps(obj))
return
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [reply joblib/test/test_hashing.py:495]: Actually not, it's probably just memoization of the top level object. I think this is kind of expected, no? I am not sure we should do anything about it.

File: joblib/test/test_hashing.py
Comment (pierreglaser) [reply joblib/test/test_hashing.py:492]: Literrally the current master branch. I copy-pasted this test into master and it did not pass.

File: joblib/hashing.py
Comment (pierreglaser) [reply joblib/hashing.py:236]: To go to the bottom of this we should look exactly at what `copy.deepcopy` does.

File: CHANGES.rst
Comment (pierreglaser) [reply CHANGES.rst:9]: You're right, thanks.

File: doc/memory.rst
Comment (pierreglaser) [reply doc/memory.rst:403]: It is true that as of now, the tests I have written would not ensure such a contract is respected. I'm still a bit un-easy about the fact that upgrading `joblib` to for example pull a bugfix (something that may be necesary), can make user risk cache invalidation.

Maybe a intermediate would be to state best-effort hash stability across `joblib` releases.

But I'll leave it to you to have the last word on this.

File: joblib/hashing.py
Comment (pierreglaser) [reply joblib/hashing.py:237]: The logic deserves a comment, but LGTM in principle.

File: joblib/test/test_hashing.py
Comment (pierreglaser) [reply joblib/test/test_hashing.py:495]: > I think this is kind of expected, no? 

My guding principle about hashing is that only value and not identity should play a role in hash values - so personally I would call this behavior a bug. Not sure we should fix this in this PR though.

File: joblib/test/test_hashing.py
Comment (pierreglaser) [reply joblib/test/test_hashing.py:480]: good call.

File: joblib/hashing.py
Comment (pierreglaser) [reply joblib/hashing.py:237]: Tests pass, and I think this method is even clearer than the previous one (see the comment I'll write)

File: joblib/test/test_hashing.py
Comment (pierreglaser) [reply joblib/test/test_hashing.py]: Unless I'm mistaken, what you propose and what' I am doing is the same thing apart form the fact that you use two executors (which is safer since you know which task is going to which process) and me one only.
So I'm replacing my lines with your suggestion.

File: doc/memory.rst
Comment (ogrisel) [reply doc/memory.rst:403]: > Maybe a intermediate would be to state best-effort hash stability across joblib releases. 

I would rather say that most joblib upgrades are unlikely to result in a change of the hashing results but this is not guaranteed and can happen from time to time in particular when fixing bugs.

File: joblib/test/test_hashing.py
Comment (ogrisel) [reply joblib/test/test_hashing.py]: The lists objects are not the same because one has 2 refs to the same object and this is not the case for the other. If we had nested graph structures with cyclic references, e.g. to represent data graphs, having 2 different hashes when only the reference structure changes, this would make sense to me.

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:485]: I would just remove those commented lines.

File: CHANGES.rst
Comment (pierreglaser) [inline CHANGES.rst:8]: I should include "joblib hash values are now not guaranteed to stay the same across joblib versions". Waiting for CI to finish before adding this.

File: joblib/hashing.py
Comment (ogrisel) [inline joblib/hashing.py:237]:
```suggestion
            # To prevent memoization from interfering with hashing, we isolate
            # the serialization (and thus the pickle memoization) of each dtype
            # using each time a different ``pickle.dumps`` call unrelated to
            # the current Hasher instance.
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:331]:
```suggestion
    # numpy dtype hashing is tricky to get right: see #231, #239, #251 #1080,
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:360]:
```suggestion
    complex_dt2 = np.dtype(
        [('name', np.str_, 16), ('grades', np.float64, (2,))]
    )

    # complex dtypes objects are not interned
    assert hash(complex_dt1) == hash(complex_dt2)
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:420]:
```suggestion
    # details of numpy to hash np.dtype objects, which makes the stability of
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:422]:
```suggestion
    # As a result, hashing stability across joblib versions becomes best-effort
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:427]:
```suggestion
    #   value. This should be viewed as a proxy for testing hash consistency
```

File: joblib/test/test_hashing.py
Comment (ogrisel) [inline joblib/test/test_hashing.py:451]: "oekoke oekoefkhnwoaersifh  oeiqo"?

Comment (ogrisel) [inline joblib/test/test_hashing.py:451]:
```suggestion
            # Non regression cases for
            # https://github.com/joblib/joblib/issues/308
```

**Review by ogrisel — APPROVED** [review]
LGTM! Thanks very much @pierreglaser!

File: doc/memory.rst
Comment (pierreglaser) [inline doc/memory.rst:394]:
```suggestion
  that are constructs (or contain references to contructs) defined in the
```

File: joblib/test/test_hashing.py
Comment (pierreglaser) [inline joblib/test/test_hashing.py:335]:
```suggestion
    # In this test, we make sure that the pickling of numpy dtypes is robust
    # to
```

**Review by pierreglaser — COMMENTED** [review]
(last round of typo correction)

File: joblib/test/test_hashing.py
Comment (pierreglaser) [reply joblib/test/test_hashing.py]: don't you just oekoke oekoefkhnwoaersifh oeiqo sometimes?

**Comment by ogrisel** [comment]
Merged!

**Comment by tomMoral** [comment]
I did not follow all the discussion but it looked like a complicated one to get right!
Thanks a lot to you both :)

**Pull request merged**
